File: src/receipts.js
'use strict';
// Local, tamper-evident audit receipts — one JSON object per line in
// .certior/audit/YYYY-MM-DD.jsonl. The log is a hash chain: each receipt carries
// seq, prev (the previous receipt's hash), and hash (SHA-256 over its own
// content). Altering or deleting any receipt breaks every later link, so
// `certior-guard verify` can prove the log is intact. Best-effort throughout:
// never throws into the hook.
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

const AUDIT_DIRNAME = path.join('.certior', 'audit');
const GENESIS = 'sha256:genesis';

// Deterministic JSON with recursively sorted keys, for stable hashing.
function stableStringify(obj) {
  if (obj === null || typeof obj !== 'object') return JSON.stringify(obj);
  if (Array.isArray(obj)) return `[${obj.map(stableStringify).join(',')}]`;
  const keys = Object.keys(obj).sort();
  return `{${keys.map((k) => `${JSON.stringify(k)}:${stableStringify(obj[k])}`).join(',')}}`;
}

function policyHash(profileDict, mode) {
  const blob = stableStringify({ profile: profileDict, mode });
  return 'sha256:' + crypto.createHash('sha256').update(blob).digest('hex').slice(0, 16);
}

// SHA-256 over a receipt's canonical content (every field except `hash`).
function receiptHash(receipt) {
  const content = { ...receipt };
  delete content.hash;
  return 'sha256:' + crypto.createHash('sha256').update(stableStringify(content)).digest('hex');
}

function nowIso() {
  return new Date().toISOString().replace(/\.\d+Z$/, 'Z');
}

function dayFiles(auditDir) {
  let entries;
  try { entries = fs.readdirSync(auditDir); } catch { return []; }
  return entries.filter((f) => f.endsWith('.jsonl')).sort().map((f) => path.join(auditDir, f));
}

// Best-effort exclusive lock via an O_EXCL lockfile; spins briefly, then proceeds.
function withLock(auditDir, fn) {
  const lockpath = path.join(auditDir, '.lock');
  let fd = -1;
  for (let attempt = 0; attempt < 50; attempt++) {
    try { fd = fs.openSync(lockpath, 'wx'); break; } catch { /* held — retry */ }
    const until = Date.now() + 5;
    while (Date.now() < until) { /* tiny spin */ }
  }
  try { return fn(); } finally {
    if (fd !== -1) { try { fs.closeSync(fd); } catch {} }
    try { fs.unlinkSync(lockpath); } catch {}
  }
}

// [seq, hash] of the last receipt in the chain, or [0, GENESIS].
function tail(auditDir) {
  const files = dayFiles(auditDir);
  for (let i = files.length - 1; i >= 0; i--) {
    let last = null;
    try {
      for (const line of fs.readFileSync(files[i], 'utf8').split('\n')) if (line.trim()) last = line;
    } catch { continue; }
    if (last) {
      try { const rec = JSON.parse(last); return [Number(rec.seq || 0), String(rec.hash || GENESIS)]; } catch { return [0, GENESIS]; }
    }
  }
  return [0, GENESIS];
}

function writeReceipt(opts) {
  try {
    const { auditDir } = opts;
    fs.mkdirSync(auditDir, { recursive: true });
    return withLock(auditDir, () => {
      const [prevSeq, prevHash] = tail(auditDir);
      const ts = nowIso();
      const receipt = {
        seq: prevSeq + 1,
        prev: prevHash,
        actor: 'claude-code',
        tool: opts.tool,
        target: String(opts.target || '').slice(0, 500),
        decision: opts.decision,
        would: opts.would,
        capability: opts.capability,
        reason: opts.reason,
        profile: opts.profileKey,
        mode: opts.mode,
        policy_hash: opts.policyHash,
        session_id: opts.sessionId || '',
        timestamp: ts,
        verifier: 'certior-guard',
      };
      receipt.hash = receiptHash(receipt);
      const file = path.join(auditDir, `${ts.slice(0, 10)}.jsonl`);
      fs.appendFileSync(file, JSON.stringify(receipt) + '\n');
      return file;
    });
  } catch { return null; }
}

function readAll(auditDir) {
  const rows = [];
  for (const file of dayFiles(auditDir)) {
    try {
      for (const line of fs.readFileSync(file, 'utf8').split('\n')) {
        const t = line.trim();
        if (t) rows.push(JSON.parse(t));
      }
    } catch { /* skip unreadable/partial */ }
  }
  return rows;
}

function readRecent(auditDir, limit = 20) { return readAll(auditDir).slice(-limit); }

module.exports = {
  AUDIT_DIRNAME, GENESIS, stableStringify, policyHash, receiptHash,
  writeReceipt, readAll, readRecent,
};

File: certior_guard/receipts.py
"""Local audit receipts — every decision is logged, no cloud, no login.

One JSON object per line, appended to ``.certior/audit/YYYY-MM-DD.jsonl``: a
grep-able, diff-able, offline record of what the agent tried and what Certior did.

A ``policy_hash`` binds each receipt to the exact rule set in force, so a receipt
can later be replayed and checked against the policy that produced it.
"""
from __future__ import annotations

import hashlib
import json
import os
from datetime import datetime, timezone
from typing import Dict, List, Optional

AUDIT_DIRNAME = os.path.join(".certior", "audit")


def policy_hash(profile_dict: Dict[str, object], mode: str) -> str:
    blob = json.dumps({"profile": profile_dict, "mode": mode}, sort_keys=True)
    return "sha256:" + hashlib.sha256(blob.encode()).hexdigest()[:16]


def _now() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def write_receipt(
    *,
    audit_dir: str,
    tool: str,
    target: str,
    decision: str,
    would: str,
    capability: str,
    reason: str,
    profile_key: str,
    mode: str,
    policy_hash: str,
    session_id: str = "",
) -> Optional[str]:
    """Append one receipt line. Best-effort — never raises into the hook."""
    try:
        os.makedirs(audit_dir, exist_ok=True)
        ts = _now()
        receipt = {
            "actor": "claude-code",
            "tool": tool,
            "target": target[:500],
            "decision": decision,
            "would": would,
            "capability": capability,
            "reason": reason,
            "profile": profile_key,
            "mode": mode,
            "policy_hash": policy_hash,
            "session_id": session_id,
            "timestamp": ts,
            "verifier": "certior-guard",
        }
        day = ts[:10]
        path = os.path.join(audit_dir, f"{day}.jsonl")
        with open(path, "a", encoding="utf-8") as fh:
            fh.write(json.dumps(receipt, ensure_ascii=False) + "\n")
        return path
    except Exception:
        return None


def read_recent(audit_dir: str, limit: int = 20) -> List[Dict[str, object]]:
    """Most recent receipts across all day-files, newest last."""
    rows: List[Dict[str, object]] = []
    try:
        files = sorted(
            os.path.join(audit_dir, f) for f in os.listdir(audit_dir) if f.endswith(".jsonl")
        )
    except FileNotFoundError:
        return []
    for path in files:
        try:
            with open(path, encoding="utf-8") as fh:
                for line in fh:
                    line = line.strip()
                    if line:
                        rows.append(json.loads(line))
        except Exception:
            continue
    return rows[-limit:]
